libprof.1
.\"	@(#) a64l.so.1 
.\"
.TH libprof.so.1 1 
.AT 3
.SH NAME
libprof.so.1 \- shared library to profile functions in libc

.SH SYNOPSIS
.B LD_PRELOAD=./libprof.so.1 
program [args] ...
.sp 0
.SH DESCRIPTION
.I libprof
intercepts calls to libc and reports their cpu-time usage.
.PP
This is used to identify hot-spots in library usage by a program,
like the 
.I prof
source-code profiler. Unlike the source profiler, libprof doesn't
require recompilation, but works on unmodified binaries.
.PP

When loaded with a program, it emits a time profile, such as:
.nj
.nf
Name                           Duration   Pct.        Calls Sec/Call
----                           ---------- ----        ----- ----------
Entire program                 0.23329523 100.0           1 0.23329523
a64l                           0.11952140 51.23      100000 0.00000120
addseverity                    0.01279909  5.56           1 0.01279909
Measurement overhead           0.00031536  0.14           1 0.00031536

.ju
.fi

.PP
This is sorted from largest to smallest duration, always beginning with
an entry for the whole program. One of the entries reports on the
(measurement) overhead of the libprof library.
.PP
If the program profiled is a shell script, one profile per binary
program is emitted, labelled with the pid of the program. 
.I ptree
(1) can be used to clarify which pid(s) are of interest.

.SH FILES
.sp 0

.SH "SEE ALSO"
truss -c (1), prof(1).

.SH BUGS
Not thread-aware, so that two threads entering the same 
function at once can clash. On Solaris this causes an
assertion failure and the program being tested to
hang in lwp_park. This is known to apply to
.nf
.nj
	mozilla
	acroread 5
	gij
	intop
	ntop
	rmic
	svn
	wish
.fi
.ju
.for example, pstack shows mozilla hung in lwp_park, thusly
.nf
.nj
698:	/usr/local/mozilla/mozilla-bin
 ff2c5f88 lwp_park (0, 0, 0)
 ff2c1d00 mutex_lock_queue (ff2d8b84, 0, ff23fbc8, ff2d8000, ff3d90d4, ff3ec7ac) + 104
 ff2c2700 slow_lock (ff23fbc8, fed30000, 0, 0, ff3c4ed8, 0) + 58
 ff1faeec dgettext (ff227e30, ff227e40, ff3ef288, ff23c000, 1, 0) + 24
 ff1b5ed4 _assert  (ff376648, ff376658, 1da, ff216b4c, 0, 0) + 18
 ff324bb0 calloc   (1, 18, 0, 0, 0, 0) + b8
 ff1faf08 dgettext (ff227e30, ff227e40, 0, ff23c000, 0, 0) + 40
 ff1b5ed4 _assert  (ff377928, ff377938, b52, 0, 0, 0) + 18
 ff3349ec getenv   (ff2c6570, 1e454, 0, 8, ff3bb7b8, 0) + b4
 ff2b9bc0 envvar   (ff2c6570, f4240, 0, fffffff8, 0, ffbfe791) + 18
 ff2b9ed4 _lib_init (ff3ed7b8, 0, ff3ec204, 3127c, ff3c3068, 0) + 290
 ff2c6270 _init    (ff3ec0c0, ff2a1cec, 18, fe801548, 2, ff3ec7ac) + 4
 ff3bf228 call_init (fe801f70, 0, fe801f70, ffdfffff, 400000, 80000) + 1c0
 ff3be7fc setup    (ffbfee5c, ff3ec194, 10000, 0, ff3ec000, 0) + 135c
 ff3cc400 _setup   (e265, a, a, ffbfffce, ffbfffda, ff3cc03c) + 3a8
 ff3b3994 _rt_boot (0, 0, 0, 0, 0, 0) + 88
 00000000 ???????? (0, 0, 0, 0, 0, 0)
.fi
.ju
.PP
Doesn't allow turning off headers. Doesn't allow sorting by 
calls column (use sort -k 4rn) or by seconds/call
(use sort -k 5rn).
.PP
Cannot report interim results in sorted order (sorting invalidates
further data collection).


libprof.c
/*
 * libprof -- intercept calls to libc and report profiling information
 *
 *   Copyright (C) 2005 David Collier-Brown, Michael B.Allen
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *   
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *   
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>

#include "libprof.h"


/* Interposer variables, implemented as file statics. */
static void printProfile(int);
static double summarize(struct prof_t *p);
static char *getProgName();
static void sigHandler(int s, siginfo_t *reason, void *context);

/* Table of stats, imported from .i files. */
extern size_t ProfCount;
extern struct prof_t ProfStats[];
#define OVERHEAD 0
#define REST ProfCount+1

/* Misc. */
#define UNSORTED 0
#define SORTED 1

#ifdef __sun
extern int gettimeofday(struct timeval *, void *);
#endif
#ifdef __linux
extern int gettimeofday(struct timeval *, __timezone_ptr_t);
#define MAXLINE 1024
#endif

static double StartTime = 0.0;

#define RTLD_SETTING RTLD_NOW /* RTLD_LAZY for production, _NOW for dev't. */

#ifdef linux
#define RTLD_NEXT ((void *) -1l)
#endif

/*
 * profInit -- get start time and set upsignal handlers for
 *	SIGTERM and SIGUSR1. SIGTERM is the traditional ^C, 
 *	and terminates the program after printing the stats
 *	in soretd order.
 */
 void 
profInit() {
	struct sigaction newAct, oldAct;
	struct timeval t;

	/* Set up signal handler for SIGTERM */
	newAct.sa_sigaction = sigHandler;
	(void) sigemptyset(&newAct.sa_mask);
	/* newAct.sa_flags = SA_SIGINFO|SA_RESTART; */
	if (sigaction(SIGTERM, &newAct, &oldAct) == -1) {
		(void) fprintf(stderr, "libprof: sigaction(SIGTERM) failed\n");
		(void) _exit(-1);
	}
	if (sigaction(SIGUSR1, &newAct, &oldAct) == -1) {
		(void) fprintf(stderr, "libprof: sigaction(SIGTUSR1) failed\n");
		(void) _exit(-1);
	}

	/* And start data collection. */
	(void) gettimeofday(&t, NULL);
	StartTime = (double) (t.tv_sec + (t.tv_usec/1000000.0)); 
}

/*
 * profFini -- print out final results. 
 */
 void
profFini() {
	printProfile(SORTED);
}

/*
 * sigHandler -- print profiling information, and optionally exit.
 */
 /*ARGSUSED*/
 static void
sigHandler(int s, siginfo_t *reason, void *context) {

	if (s == SIGUSR1) {
		/* Don't sort until done collecting. */
	        printProfile(UNSORTED);
	}
	else {
		printProfile(SORTED);
	}
	/* Return zero, as we succeeded in profiling. */
	/* Odd, but that's what truss -c does, so... */
	(void) exit(0);
}
     

/*
 * compare -- compare two prof_t's, for qsort in printProfile
 */
static int compare(const void *this, const void *that) {
	double d;

	d = (((struct prof_t *)that)->time - 
		((struct prof_t *)this)->time);
	if (d > 0.00000001) {
		return 1;
	}
	else if (d < 0) {
		return -1;
	}
	else {
		return 0;
	}
}

/*
 * summarize -- sum up a table of struct prof_t *s 
 */
 static double 
summarize(struct prof_t *p) {
	double sum = 0.0;

	while (p->name != NULL) {
		sum += p->time;
		p++;
	}
	return sum;
}
	
/*
 * printProfile -- print the results
  */
 static void
printProfile(int sort) {
	struct prof_t	*p;
	struct timeval t;
	double	totalTime, programTime;
	double	overhead = 0.0;
	static void *myprintf = NULL;

	/* Initialize. */
	(void) gettimeofday(&t, NULL);
	totalTime = (double) (t.tv_sec + (t.tv_usec/1000000.0)) - StartTime; 
	myprintf = dlsym(RTLD_NEXT, "fprintf");
	assert(myprintf != NULL);
#define MYFPRINTF ((int (*)(FILE *, char *, ...))myprintf)

	/* Print the header. */
	(void) MYFPRINTF(stderr,"\nResource usage profile for "
		"process %ld, %s\n", getpid(), getProgName());
	(void) MYFPRINTF(stderr, 
		"%-30s %15s %-5s %11s %-10s\n",
		 "Name", "Duration", "Pct.", "Calls", "Sec/Call");

	(void) MYFPRINTF(stderr,
		"%-30s %15s %5s %11s %10s\n",
		"----", "------------", "-----", "-----", "----------");
	/* The time for the whole program. */
	(void) MYFPRINTF(stderr, 
		"%-30.30s %15.8f %4.1f %11ld %.8f\n",
		"Entire program",
		totalTime,
		(double) 100.0,
		(long) 1,
		totalTime);

	/* Add overhead at ~0.00002628 seconds per interface */
	/* for version using hrtime... */
#ifdef USING_HRTIME
	for (p=&ProfStats[0]; p->name != NULL; p++) {
		if (p->calls > 0) {
			overhead += 0.00002628;
		}
	}
	ProfStats[OVERHEAD].calls = 1;
	ProfStats[OVERHEAD].time = overhead * 1.0E9;
#else
	overhead = 0.00002628;
	ProfStats[OVERHEAD].calls = 1;
	ProfStats[OVERHEAD].time = overhead;
#endif

	/* print time for the non-library parts of the program. */
	programTime = totalTime - summarize(&ProfStats[0]);
	(void) MYFPRINTF(stderr, 
		"%-30.30s %15.8f %5.2f %11ld %.8f\n",
		"Non-library time",
		programTime,
		(double) programTime * 100.0 / totalTime,
		(long) 1,
		programTime);
	(void) MYFPRINTF(stderr,"%-30s\n","----");

	/* Sort the table, which makes further collection invalid. */
	if (sort == SORTED) {
		qsort(ProfStats, ProfCount, sizeof(struct prof_t), compare);
	}

	/* Print it out. */
	for (p=&ProfStats[0]; p->name != NULL; p++) {
		if (p->calls > 0) {
			(void) MYFPRINTF(stderr, 
				"%-30.30s %15.8f %5.2f %11ld %.8f\n",
				p->name,
				p->time, /* Time, seconds */
				(p->time * 100.0) / totalTime,
				p->calls, 
				p->time / p->calls);
		}
	}
}



/*
 * getProgName -- get the name from a struct which contains it.
 */
#if defined(__sun)
 char *
getProgName() {
        int     fdp;
        static struct psinfo pbuff;

	if ((fdp = open("/proc/self/psinfo", O_RDONLY, 0)) < 0) {
		return NULL;
	}	
	if (read(fdp, (void *)&pbuff, sizeof(pbuff)) != (int)sizeof(pbuff)) {
		(void) close(fdp);
		return NULL;
	}
	(void) close(fdp);
	return pbuff.pr_fname;
}


#elif defined(__linux)

static char *skipPrefix(char *);

 char *
getProgName() {
	int     fdp;
        static char line[MAXLINE];

	if ((fdp = open("/proc/self/cmdline", 0 /* O_RDONLY */,0)) < 0) {
		return NULL;
	}	
	if (read(fdp, (void *)&line, sizeof(line)) <= 0) {
		(void) close(fdp);
		return NULL;
	}
	(void) close(fdp);
	return skipPrefix(&line[0]);
}

/*
 * skipPrefix -- get to the basename part of a path
 */
 static char *
skipPrefix(char *p) {
	char	*q;

	if ((q = strrchr(p, '/')) != NULL) {
		return q+1;
	}
	else {
		return p;	
	}
}

#ifdef GETHRTIME
/*
 * gethrtime -- hi-res timer for Linux, courtesy of 
 *	Sternberg. Returns time in nanoseconds, aka 
 *	cycles/1GHz
 */
unsigned long gethrtime(void) {
	static unsigned long cpuSpeed = 0;
	unsigned long result;

	if (cpuSpeed == 0) {
		cpuSpeed = GetCpuSpeed();
	}

	asm("rdtsc":"=A" (result));
	return (unsigned long) (result/cpuSpeed);
}

/*
 * GetCpuSpeed -- get cpu speed, courtesy of Michael B Allen.
 * Returns CPU clock in khz, from playstation linux port.
 */
static unsigned int GetCpuSpeed(void) {
	unsigned long long int tscstart, tscstop;
	unsigned int start, stop;

	tscstart = rdtsc();
	start = GetTimer();
	usec_sleep(50000);
	stop = GetTimer();
	tscstop = rdtsc();

	return((tscstop-tscstart)/((stop-start)/1000.0));
}
#endif /* GETHRTIME */

#endif /* linux */
